packages/repository/src/relations/has-many/has-many-through-repository.factory.ts

```diff
@@ -0,0 +1,87 @@
+
+import * as debugFactory from 'debug';
+import {DataObject} from '../../common-types';
+import {Entity} from '../../model';
+import {EntityCrudRepository} from '../../repositories/repository';
+import {Getter, HasManyThroughDefinition} from '../relation.types';
+import {
+  createTargetConstraint,
+  createThroughConstraint,
+  resolveHasManyThroughMetadata,
+} from './has-many-through.helpers';
+import {
+  DefaultHasManyThroughRepository,
+  HasManyThroughRepository,
+} from './has-many-through.repository';
+
+const debug = debugFactory(
+  'loopback:repository:has-many-through-repository-factory',
+);
+
+export type HasManyThroughRepositoryFactory<
+  Target extends Entity,
+  Through extends Entity,
+  ForeignKeyType
+> = (fkValue: ForeignKeyType) => HasManyThroughRepository<Target, Through>;
+
+/**
+ * Enforces a constraint on a repository based on a relationship contract
+ * between models. For example, if a Customer model is related to an Order model
+ * via a HasMany relation, then, the relational repository returned by the
+ * factory function would be constrained by a Customer model instance's id(s).
+ *
+ * @param relationMetadata - The relation metadata used to describe the
+ * relationship and determine how to apply the constraint.
+ * @param targetRepositoryGetter - The repository which represents the target model of a
+ * relation attached to a datasource.
+ * @returns The factory function which accepts a foreign key value to constrain
+ * the given target repository
+ */
+export function createHasManyThroughRepositoryFactory<
+  Target extends Entity,
+  TargetID,
+  Through extends Entity,
+  ThroughID,
+  ForeignKeyType
+>(
+  relationMetadata: HasManyThroughDefinition,
+  targetRepositoryGetter: Getter<EntityCrudRepository<Target, TargetID>>,
+  throughRepositoryGetter: Getter<EntityCrudRepository<Through, ThroughID>>,
+): HasManyThroughRepositoryFactory<Target, Through, ForeignKeyType> {
+  const meta = resolveHasManyThroughMetadata(relationMetadata);
+  debug('Resolved HasManyThrough relation metadata: %o', meta);
+  return function(fkValue?: ForeignKeyType) {
+    function getTargetContraint(
+      throughInstances: Through[],
+    ): DataObject<Target> {
+      return createTargetConstraint<Target, Through>(meta, throughInstances);
+    }
+    function getThroughConstraint(
+      targetInstance?: Target,
+    ): DataObject<Through> {
+      const constriant: DataObject<Through> = createThroughConstraint<
+        Target,
+        Through,
+        ForeignKeyType
+      >(meta, fkValue, targetInstance);
+      return constriant;
+    }
+    return new DefaultHasManyThroughRepository<
+      Target,
+      TargetID,
+      EntityCrudRepository<Target, TargetID>,
+      Through,
+      ThroughID,
+      EntityCrudRepository<Through, ThroughID>
+    >(
+      targetRepositoryGetter,
+      throughRepositoryGetter,
+      getTargetContraint,
+      getThroughConstraint,
```

Comment:
Re-starting the thread https://github.com/strongloop/loopback-next/pull/2359#discussion_r257276433 here.

> Take a look at my following comment. Perhaps it is best to come back to this later.
> I suggest adding the PK/FK constraints in a separate pull request. I created a feature request for it here #3545.

My problem with this proposal is that the way how you designed the API, #3545 will require breaking changes and thus a new semver-major release.

My comment https://github.com/strongloop/loopback-next/pull/2359#discussion_r257276433 includes a proposal that will be forward-compatible with SQL-JOIN-based implementation and can be implemented using multiple DB queries too, but I understand you may be reluctant to follow that proposal.

I feel we have to make a trade-off here:
1. We can accept the current proposal and (hopefully) get HasManyThrough relation available sooner, at the future expense of breaking the API & releasing a semver-major version when we get to implement SQL JOINs.
2. Alternatively, we can insist on a forward-compatible API at the cost of delaying the delivery of HasManyThrough relation.

Considering that SQL JOINs are not on our roadmap yet, I think we won't get to implement SQL JOINs in the next 6 months, thus it should be ok to bump the major version when we eventually get to work on SQL JOINs. Unless there is a community contribution for this feature coming sooner.

I'd like to hear from @strongloop/loopback-maintainers, especially @raymondfeng and @dhmlau: what is your opinion?

Reply:
Just my 2 cents.

It's tempting to have HasManyThrough relation available sooner, given the number of our users have indicated they want this feature back, see https://github.com/strongloop/loopback-next/issues/2043. I think if we won't need to have a semver-major version released in the next 6 months (at least) because of this, it's reasonable.

Also, @raymondfeng @bajtos, I remember we were once discussing there are some breaking changes that we might need to make too, though I couldn't recall which area or feature. 

Would definitely like to hear opinions from @raymondfeng and @strongloop/loopback-maintainer because I'm not as close to the implementation as the rest. :)